File: src/main/java/voidchess/board/ChessGameInterface.java
package voidchess.board;

import voidchess.helper.ChessGameSupervisor;
import voidchess.helper.Move;
import voidchess.helper.Position;

import java.util.List;

/**
 * @author stephan
 */
public interface ChessGameInterface extends BasicChessGameInterface {
    int NO_END = 0;
    int MATT = 1;
    int PATT = 2;
    int DRAW = 3;
    int THREE_TIMES_SAME_POSITION = 4;
    int FIFTY_MOVES_NO_HIT = 5;
    int RESIGN = 6;

    void initGame(int chess960);

    /**
     * returns true if this game was started with the standard distribution of figures
     * (chess960 code: 518)
     */
    boolean isStandardGame();

    List<ChessGameInterface> copyGame(int neededInstances);

    boolean isWhiteTurn();

    int countFigures();

    boolean hasHitFigure();

    boolean isCheck(boolean isWhite);

    boolean isSelectable(Position pos, boolean whitePlayer);

    boolean isMoveable(Position from, Position to, boolean whitePlayer);

    boolean isFreeArea(Position pos);

    int move(Move move);

    void getPossibleMoves(List<Move> possibleMoves);

    int countReachableMoves(boolean isWhite);

    ChessGameSupervisor suspendInteractiveSupervisor();

    void useSupervisor(ChessGameSupervisor supervisor);

    void undo();

    /**
     * @return all the moves played so far in a string representation
     */
    String getCompleteHistory();
}

File: src/main/kotlin/voidchess/board/ChessGameInterface.kt
package voidchess.board

import voidchess.board.move.ExtendedMove
import voidchess.board.move.Move
import voidchess.board.move.MoveResult
import voidchess.board.move.Position


interface ChessGameInterface : BasicChessGameInterface {

    /**
     * returns true if this game was started with the standard distribution of figures
     * (chess960 code: 518)
     */
    val isStandardGame: Boolean
    val isWhiteTurn: Boolean

    /**
     * @return all the moves played so far in a string representation
     */
    fun getCompleteHistory(): String
    fun getLastExtendedMove(): ExtendedMove
    fun initGame(chess960: Int)
    fun copyGame(neededInstances: Int): List<ChessGameInterface>
    fun countFigures(): Int
    fun hasHitFigure(): Boolean
    fun isCheck(isWhiteInCheck: Boolean): Boolean
    fun isSelectable(pos: Position, whitePlayer: Boolean): Boolean
    fun isMovable(from: Position, to: Position, whitePlayer: Boolean): Boolean
    fun move(move: Move): MoveResult
    fun getPossibleMoves(possibleMoves: MutableList<Move>)
    /**
     * @return a pair of ints with the first int denoting the reachable sum of moves of all white figures
     * while the second int is the same for all black figures.
     */
    fun countReachableMoves(): Pair<Int, Int>
    fun suspendInteractiveSupervisor(): ChessGameSupervisor
    fun useSupervisor(supervisor: ChessGameSupervisor)
    fun undo()
}